### ISSUE TITLE
Different result of evaluation

### ISSUE THREAD

**Issue by Kryptonites** [google-research/bert#399]
When I use the same evaluation set to run "do eval"without "do_train" and "do_train,do_eval", there is two different result presented, former is much lower than later.
Is there any wrong setting?

**Comment by mat7213** [comment]
I have the same question...
Train an eval：
`python run_classifier.py \
--data_dir=$DATA_DIR \
--task_name=xxx \
--do_train=true \
--do_eval=true \
--do_predict=false \
--vocab_file=$BERT_BASE_DIR/vocab.txt \
--bert_config_file=$BERT_BASE_DIR/bert_config.json \
--init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt \
--max_seq_length=32 \
--train_batch_size=32 \
--learning_rate=2e-5 \
--num_train_epochs=3.0 \
--output_dir=../output`

got the result：
`...
INFO:tensorflow:Saving checkpoints for 306482 into .../output/model.ckpt.
INFO:tensorflow:Loss for final step: 0.68741745.
INFO:tensorflow:training_loop marked as finished
INFO:tensorflow:Writing example 0 of 181618
INFO:tensorflow:*** Example ***
...
INFO:tensorflow:Finished evaluation at 2019-03-05-17:12:46
INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.7659318, eval_loss = 1.156508, global_step = 306482, loss = 1.1564742
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
INFO:tensorflow:evaluation_loop marked as finished
INFO:tensorflow:***** Eval results *****
INFO:tensorflow:  eval_accuracy = 0.7659318
INFO:tensorflow:  eval_loss = 1.156508
INFO:tensorflow:  global_step = 306482
INFO:tensorflow:  loss = 1.1564742`


after that , i run 
`python run_classifier.py \
--data_dir=$DATA_DIR \
--task_name=xxx \
--do_eval=true \
--do_predict=false \
--vocab_file=$BERT_BASE_DIR/vocab.txt \
--bert_config_file=$BERT_BASE_DIR/bert_config.json \
--init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt \
--max_seq_length=32 \
--train_batch_size=32 \
--learning_rate=2e-5 \
--num_train_epochs=3.0 \
--output_dir=../output`

got different result:
`INFO:tensorflow:Finished evaluation at 2019-03-06-02:57:04
INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.0013820216, eval_loss = 8.9281435, global_step = 306482, loss = 8.928147
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
INFO:tensorflow:evaluation_loop marked as finished
INFO:tensorflow:***** Eval results *****
INFO:tensorflow:  eval_accuracy = 0.0013820216
INFO:tensorflow:  eval_loss = 8.9281435
INFO:tensorflow:  global_step = 306482
INFO:tensorflow:  loss = 8.928147`

**Comment by Kryptonites** [comment]
> I have the same question...
> Train and eval：
> 
> > python run_classifier.py 
> > --data_dir=$DATA_DIR 
> > --task_name=xxx 
> > --do_train=true 
> > --do_eval=true 
> > --do_predict=false 
> > --vocab_file=$BERT_BASE_DIR/vocab.txt 
> > --bert_config_file=$BERT_BASE_DIR/bert_config.json 
> > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt 
> > --max_seq_length=32 
> > --train_batch_size=32 
> > --learning_rate=2e-5 
> > --num_train_epochs=3.0 
> > --output_dir=../output
> 
> got the result：
> 
> > ...
> > INFO:tensorflow:Saving checkpoints for 306482 into .../output/model.ckpt.
> > INFO:tensorflow:Loss for final step: 0.68741745.
> > INFO:tensorflow:training_loop marked as finished
> > INFO:tensorflow:Writing example 0 of 181618
> > INFO:tensorflow:*** Example ***
> > ...
> > INFO:tensorflow:Finished evaluation at 2019-03-05-17:12:46
> > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.7659318, eval_loss = 1.156508, global_step = 306482, loss = 1.1564742
> > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > INFO:tensorflow:evaluation_loop marked as finished
> > INFO:tensorflow:***** Eval results *****
> > INFO:tensorflow:  eval_accuracy = 0.7659318
> > INFO:tensorflow:  eval_loss = 1.156508
> > INFO:tensorflow:  global_step = 306482
> > INFO:tensorflow:  loss = 1.1564742
> 
> after that , i run
> 
> > python run_classifier.py 
> > --data_dir=$DATA_DIR 
> > --task_name=xxx 
> > --do_eval=true 
> > --do_predict=false 
> > --vocab_file=$BERT_BASE_DIR/vocab.txt 
> > --bert_config_file=$BERT_BASE_DIR/bert_config.json 
> > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt 
> > --max_seq_length=32 
> > --train_batch_size=32 
> > --learning_rate=2e-5 
> > --num_train_epochs=3.0 
> > --output_dir=../output
> 
> got different result:
> 
> > INFO:tensorflow:Finished evaluation at 2019-03-06-02:57:04
> > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.0013820216, eval_loss = 8.9281435, global_step = 306482, loss = 8.928147
> > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > INFO:tensorflow:evaluation_loop marked as finished
> > INFO:tensorflow:***** Eval results *****
> > INFO:tensorflow:  eval_accuracy = 0.0013820216
> > INFO:tensorflow:  eval_loss = 8.9281435
> > INFO:tensorflow:  global_step = 306482
> > INFO:tensorflow:  loss = 8.928147

During training, the model shuffle the training set, so that if you separate training and evaluation process apart, there would be a mismatch happen between training label list and evaluation label list.
You can save the training label list, and reusing it when you evaluate the model.

**Comment by mat7213** [comment]
> > I have the same question...
> > Train and eval：
> > > python run_classifier.py
> > > --data_dir=$DATA_DIR
> > > --task_name=xxx
> > > --do_train=true
> > > --do_eval=true
> > > --do_predict=false
> > > --vocab_file=$BERT_BASE_DIR/vocab.txt
> > > --bert_config_file=$BERT_BASE_DIR/bert_config.json
> > > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt
> > > --max_seq_length=32
> > > --train_batch_size=32
> > > --learning_rate=2e-5
> > > --num_train_epochs=3.0
> > > --output_dir=../output
> > 
> > 
> > got the result：
> > > ...
> > > INFO:tensorflow:Saving checkpoints for 306482 into .../output/model.ckpt.
> > > INFO:tensorflow:Loss for final step: 0.68741745.
> > > INFO:tensorflow:training_loop marked as finished
> > > INFO:tensorflow:Writing example 0 of 181618
> > > INFO:tensorflow:*** Example ***
> > > ...
> > > INFO:tensorflow:Finished evaluation at 2019-03-05-17:12:46
> > > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.7659318, eval_loss = 1.156508, global_step = 306482, loss = 1.1564742
> > > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > > INFO:tensorflow:evaluation_loop marked as finished
> > > INFO:tensorflow:***** Eval results *****
> > > INFO:tensorflow:  eval_accuracy = 0.7659318
> > > INFO:tensorflow:  eval_loss = 1.156508
> > > INFO:tensorflow:  global_step = 306482
> > > INFO:tensorflow:  loss = 1.1564742
> > 
> > 
> > after that , i run
> > > python run_classifier.py
> > > --data_dir=$DATA_DIR
> > > --task_name=xxx
> > > --do_eval=true
> > > --do_predict=false
> > > --vocab_file=$BERT_BASE_DIR/vocab.txt
> > > --bert_config_file=$BERT_BASE_DIR/bert_config.json
> > > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt
> > > --max_seq_length=32
> > > --train_batch_size=32
> > > --learning_rate=2e-5
> > > --num_train_epochs=3.0
> > > --output_dir=../output
> > 
> > 
> > got different result:
> > > INFO:tensorflow:Finished evaluation at 2019-03-06-02:57:04
> > > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.0013820216, eval_loss = 8.9281435, global_step = 306482, loss = 8.928147
> > > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > > INFO:tensorflow:evaluation_loop marked as finished
> > > INFO:tensorflow:***** Eval results *****
> > > INFO:tensorflow:  eval_accuracy = 0.0013820216
> > > INFO:tensorflow:  eval_loss = 8.9281435
> > > INFO:tensorflow:  global_step = 306482
> > > INFO:tensorflow:  loss = 8.928147
> 
> During training, the model shuffle the training set, so that if you separate training and evaluation process apart, there would be a mismatch happen between training label list and evaluation label list.
> You can save the training label list, and reusing it when you evaluate the model.

Thanks！
can i recover the training label list with train.tf_record ?  
It will take a lot of time to retrain...

**Comment by Kryptonites** [comment]
> > > I have the same question...
> > > Train and eval：
> > > > python run_classifier.py
> > > > --data_dir=$DATA_DIR
> > > > --task_name=xxx
> > > > --do_train=true
> > > > --do_eval=true
> > > > --do_predict=false
> > > > --vocab_file=$BERT_BASE_DIR/vocab.txt
> > > > --bert_config_file=$BERT_BASE_DIR/bert_config.json
> > > > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt
> > > > --max_seq_length=32
> > > > --train_batch_size=32
> > > > --learning_rate=2e-5
> > > > --num_train_epochs=3.0
> > > > --output_dir=../output
> > > 
> > > 
> > > got the result：
> > > > ...
> > > > INFO:tensorflow:Saving checkpoints for 306482 into .../output/model.ckpt.
> > > > INFO:tensorflow:Loss for final step: 0.68741745.
> > > > INFO:tensorflow:training_loop marked as finished
> > > > INFO:tensorflow:Writing example 0 of 181618
> > > > INFO:tensorflow:*** Example ***
> > > > ...
> > > > INFO:tensorflow:Finished evaluation at 2019-03-05-17:12:46
> > > > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.7659318, eval_loss = 1.156508, global_step = 306482, loss = 1.1564742
> > > > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > > > INFO:tensorflow:evaluation_loop marked as finished
> > > > INFO:tensorflow:***** Eval results *****
> > > > INFO:tensorflow:  eval_accuracy = 0.7659318
> > > > INFO:tensorflow:  eval_loss = 1.156508
> > > > INFO:tensorflow:  global_step = 306482
> > > > INFO:tensorflow:  loss = 1.1564742
> > > 
> > > 
> > > after that , i run
> > > > python run_classifier.py
> > > > --data_dir=$DATA_DIR
> > > > --task_name=xxx
> > > > --do_eval=true
> > > > --do_predict=false
> > > > --vocab_file=$BERT_BASE_DIR/vocab.txt
> > > > --bert_config_file=$BERT_BASE_DIR/bert_config.json
> > > > --init_checkpoint=$BERT_BASE_DIR/bert_model.ckpt
> > > > --max_seq_length=32
> > > > --train_batch_size=32
> > > > --learning_rate=2e-5
> > > > --num_train_epochs=3.0
> > > > --output_dir=../output
> > > 
> > > 
> > > got different result:
> > > > INFO:tensorflow:Finished evaluation at 2019-03-06-02:57:04
> > > > INFO:tensorflow:Saving dict for global step 306482: eval_accuracy = 0.0013820216, eval_loss = 8.9281435, global_step = 306482, loss = 8.928147
> > > > INFO:tensorflow:Saving 'checkpoint_path' summary for global step 306482: ../output/model.ckpt-306482
> > > > INFO:tensorflow:evaluation_loop marked as finished
> > > > INFO:tensorflow:***** Eval results *****
> > > > INFO:tensorflow:  eval_accuracy = 0.0013820216
> > > > INFO:tensorflow:  eval_loss = 8.9281435
> > > > INFO:tensorflow:  global_step = 306482
> > > > INFO:tensorflow:  loss = 8.928147
> > 
> > 
> > During training, the model shuffle the training set, so that if you separate training and evaluation process apart, there would be a mismatch happen between training label list and evaluation label list.
> > You can save the training label list, and reusing it when you evaluate the model.
> 
> Thanks！
> can i recover the training label list with train.tf_record ?
> It will take a lot of time to retrain...

Yes, that'll do.
